File: NSTU-lab3/lab3/src/containers/singlelist.cpp
#ifndef SINGLELIST_CPP
#define SINGLELIST_CPP

template<typename T>
SingleList<T>::Node::Node(const T& value) : data(value), next(nullptr) {}

template<typename T>
SingleList<T>::SingleList() : head(nullptr), length(0) {}

template<typename T>
SingleList<T>::~SingleList() {
    clear();
}

template<typename T>
void SingleList<T>::addHead(const T& value) {
    Node* newNode = new Node(value);
    newNode->next = head;
    head = newNode;
    length++;
}

template<typename T>
void SingleList<T>::addTail(const T& value) {
    Node* newNode = new Node(value);
    if (head == nullptr) {
        head = newNode;
    } else {
        Node* current = head;
        while (current->next != nullptr) {
            current = current->next;
        }
        current->next = newNode;
    }
    length++;
}

template<typename T>
bool SingleList<T>::remove(const T& value) {
    if (head == nullptr) return false;
    
    if (head->data == value) {
        Node* temp = head;
        head = head->next;
        delete temp;
        length--;
        return true;
    }
    
    Node* current = head;
    while (current->next != nullptr) {
        if (current->next->data == value) {
            Node* temp = current->next;
            current->next = current->next->next;
            delete temp;
            length--;
            return true;
        }
        current = current->next;
    }
    return false;
}

template<typename T>
bool SingleList<T>::contains(const T& value) const {
    Node* current = head;
    while (current != nullptr) {
        if (current->data == value) {
            return true;
        }
        current = current->next;
    }
    return false;
}

template<typename T>
void SingleList<T>::clear() {
    while (head != nullptr) {
        Node* temp = head;
        head = head->next;
        delete temp;
    }
    length = 0;
}

template<typename T>
size_t SingleList<T>::size() const {
    return length;
}

template<typename T>
bool SingleList<T>::empty() const {
    return head == nullptr;
}

template<typename T>
void SingleList<T>::print(std::ostream& os) const {
    Node* current = head;
    os << "[";
    while (current != nullptr) {
        os << current->data;
        if (current->next != nullptr) os << " -> ";
        current = current->next;
    }
    os << "]";
}

#endif

File: NSTU-lab3/lab3_go/containers/singlelist.go
package containers

import (
	"encoding/binary"
	"fmt"
	"io"
)

// SingleList - односвязный список
type SingleList struct {
	head  *singleNode
	count int
}

type singleNode struct {
	data int
	next *singleNode
}

// NewSingleList создаёт новый односвязный список
func NewSingleList() *SingleList {
	return &SingleList{
		head:  nil,
		count: 0,
	}
}

// AddHead добавляет элемент в начало
func (sl *SingleList) AddHead(value int) {
	newNode := &singleNode{data: value, next: sl.head}
	sl.head = newNode
	sl.count++
}

// AddTail добавляет элемент в конец
func (sl *SingleList) AddTail(value int) {
	newNode := &singleNode{data: value, next: nil}

	if sl.head == nil {
		sl.head = newNode
	} else {
		current := sl.head
		for current.next != nil {
			current = current.next
		}
		current.next = newNode
	}
	sl.count++
}

// Remove удаляет первое вхождение элемента
func (sl *SingleList) Remove(value int) bool {
	if sl.head == nil {
		return false
	}

	if sl.head.data == value {
		sl.head = sl.head.next
		sl.count--
		return true
	}

	current := sl.head
	for current.next != nil {
		if current.next.data == value {
			current.next = current.next.next
			sl.count--
			return true
		}
		current = current.next
	}
	return false
}

// Contains проверяет наличие элемента
func (sl *SingleList) Contains(value int) bool {
	current := sl.head
	for current != nil {
		if current.data == value {
			return true
		}
		current = current.next
	}
	return false
}

// Size возвращает количество элементов
func (sl *SingleList) Size() int {
	return sl.count
}

// Clear очищает список
func (sl *SingleList) Clear() {
	sl.head = nil
	sl.count = 0
}

// Print выводит список
func (sl *SingleList) Print() string {
	if sl.head == nil {
		return "[]"
	}

	result := "["
	current := sl.head
	first := true
	for current != nil {
		if !first {
			result += " -> "
		}
		result += fmt.Sprintf("%d", current.data)
		first = false
		current = current.next
	}
	result += "]"
	return result
}

// SaveToBinary сохраняет список в бинарном формате
func (sl *SingleList) SaveToBinary(w io.Writer) error {
	if err := binary.Write(w, binary.LittleEndian, int32(sl.count)); err != nil {
		return err
	}

	current := sl.head
	for current != nil {
		if err := binary.Write(w, binary.LittleEndian, int32(current.data)); err != nil {
			return err
		}
		current = current.next
	}
	return nil
}

// LoadFromBinary загружает список из бинарного формата
func (sl *SingleList) LoadFromBinary(r io.Reader) error {
	sl.Clear()

	var count int32
	if err := binary.Read(r, binary.LittleEndian, &count); err != nil {
		return err
	}

	for i := 0; i < int(count); i++ {
		var value int32
		if err := binary.Read(r, binary.LittleEndian, &value); err != nil {
			return err
		}
		sl.AddTail(int(value))
	}
	return nil
}
